### PR TITLE
Add a mysql function convert_tz()

### PR DESCRIPTION
**Pull request by Aristoeu**
### What changes were proposed in this pull request?
A MySQL date function "convert_tz" is added.

### Why are the changes needed?
MySQL has implemented this function, Oceanbase also need it.

### Will break the compatibility? How if so?
No, there is only one function added.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
I have attached some test cases in file expr_convert_tz.test, whose result is in expr_convert_tz.result.

### Checklist

- [x] I've run the tests to see all new and existing tests pass.
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above.
- [x] I've informed the technical writer about the documentation change if necessary.

### CODE REVIEW

File: src/sql/engine/expr/ob_expr_convert_tz.cpp [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+          ret = ret_more;
+        }
+      } else if (OB_FAIL(calc(timestamp_data, *target_tz_pos, offset))) {
+        LOG_WARN("calc failed", K(ret), K(timestamp_data));
+      } 
+		} else {
+      LOG_WARN("str to offset failed", K(ret));
+    }
+	} else {
+			LOG_INFO("str to offset succeed", K(tz_str), K(offset));
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.cpp:169]: remove or use LOG_DEBUG.
log that prints for each row should not be info level.

File: src/sql/engine/expr/ob_expr_convert_tz.cpp [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+int ObExprConvertTZ::parse_string(int64_t &timestamp_data, const ObString &tz_str, ObSQLSessionInfo *session, int32_t &offset)
+{
+	int ret = OB_SUCCESS;
+	int ret_more = 0;
+	if (OB_FAIL(ObTimeConverter::str_to_offset(tz_str, offset, ret_more,
+                              false /* oracle_mode */, true /* need_check_valid */)))
+	{
+		LOG_INFO("direct str_to_offset failed");
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.cpp:150]: use LOG_WARN if failed

File: src/sql/engine/expr/ob_expr_convert_tz.cpp [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+{
+	int ret = OB_SUCCESS;
+	int32_t offset_s = 0;
+  int32_t offset_d = 0;
+	if(OB_FAIL(ObExprConvertTZ::parse_string(timestamp_data, tz_str_s,session, offset_s))){
+		LOG_WARN("source time zone parse failed", K(tz_str_s));
+	} else if(OB_FAIL(ObExprConvertTZ::parse_string(timestamp_data, tz_str_d,session, offset_d))){
+		LOG_WARN("source time zone parse failed", K(tz_str_d));
+	} else {
+		//LOG_INFO("original timestamp", K(timestamp_data));
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.cpp:187]: remove

File: src/sql/engine/expr/ob_expr_convert_tz.cpp [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+		timestamp_data = timestamp_data + ((int64_t)(offset_d - offset_s)) * 1000000;
+		//LOG_INFO("convert tz succeed", K(timestamp_data), K(tz_str_s), K(tz_str_d), K(offset_s), K(offset_d));
+	}
+	return ret;
+}																		
+/*
+int ObExprConvertTZ::calc_convert_tz(int64_t &timestamp_data,
+                                    const ObString &tz_str_s,//源时区 input2
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.cpp:195]: remove

File: src/sql/engine/expr/ob_expr_convert_tz.cpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+	int ret = OB_SUCCESS;
+	int32_t offset_s = 0;
+  int32_t offset_d = 0;
+	if(OB_FAIL(ObExprConvertTZ::parse_string(timestamp_data, tz_str_s,session, offset_s))){
+		LOG_WARN("source time zone parse failed", K(tz_str_s));
+	} else if(OB_FAIL(ObExprConvertTZ::parse_string(timestamp_data, tz_str_d,session, offset_d))){
+		LOG_WARN("source time zone parse failed", K(tz_str_d));
+	} else {
+		//LOG_INFO("original timestamp", K(timestamp_data));
+		timestamp_data = timestamp_data + ((int64_t)(offset_d - offset_s)) * 1000000;
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.cpp:188]: use static_cast for static cast

File: src/sql/engine/expr/ob_expr_convert_tz.cpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+  int ret = OB_SUCCESS;
+  offset_sec = 0;
+  int32_t tz_id = tz_info_pos.get_tz_id();
+  int32_t tran_type_id = 0;
+  ObString tz_abbr_str;
+  if (OB_FAIL(tz_info_pos.get_timezone_offset(timestamp_data, offset_sec,
+                                              tz_abbr_str,tran_type_id))) {//返回值存在offset_sec里
+    LOG_WARN("get timezone sub offset failed", K(ret));
+  } else {
+      //offset_sec = offset_sec * 1000000;
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.cpp:257]: remove

File: src/sql/engine/expr/ob_expr_convert_tz.cpp [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+*/
+int ObExprConvertTZ::calc(int64_t &timestamp_data, const ObTimeZoneInfoPos &tz_info_pos, int32_t &offset_sec)
+{
+  int ret = OB_SUCCESS;
+  offset_sec = 0;
+  int32_t tz_id = tz_info_pos.get_tz_id();
+  int32_t tran_type_id = 0;
+  ObString tz_abbr_str;
+  if (OB_FAIL(tz_info_pos.get_timezone_offset(timestamp_data, offset_sec,
+                                              tz_abbr_str,tran_type_id))) {//返回值存在offset_sec里
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.cpp:254]: use English comment.

File: src/sql/engine/expr/ob_expr_convert_tz.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                           const common::ObObj &input2,
+                           const common::ObObj &input3,
+                           common::ObExprCtx &expr_ctx) const;
+  virtual int cg_expr(ObExprCGCtx &expr_cg_ctx, const ObRawExpr &raw_expr,
+                      ObExpr &expr) const override;
+  static int eval_convert_tz(const ObExpr &expr, ObEvalCtx &ctx, ObDatum &res);
+  static int find_time_zone_pos(const ObString &tz_name,
+                                        const ObTimeZoneInfo &tz_info,
+                                        ObTimeZoneInfoPos *&tz_info_pos);
+  static int calc_convert_tz(int64_t &timestamp_data, const ObString &tz_str_s,//源时区 input2
```
Comment (shendachuan) [inline src/sql/engine/expr/ob_expr_convert_tz.h:40]: use English comment.

**Review by LINxiansheng — APPROVED** [review]

**Comment by LINxiansheng** [comment]
Merged.
Thanks for your contribution.
https://github.com/oceanbase/oceanbase/commit/67e7063c4e469e36c45e9ab80007f464e164905d

**Pull request closed without merging**